CIP-0033/README.md

```diff
@@ -0,0 +1,117 @@
+Authors: Michael Peyton Jones <michael.peyton-jones@iohk.io>
+Comments-Summary: No comments
+Comments-URI: 
+Status: Draft
+Type: Standards Track
+Created: 2021-11-29
+License: CC-BY-4.0
+Requires: CIP-31
+---
+
+# Reference scripts
+
+## Abstract
+
+We propose to allow scripts ("reference scripts") to be attached to outputs, and to allow reference scripts to be used to satisfy script requirements during validation, rather than requiring the spending transaction to do so.
+This will allow transactions using common scripts to be much smaller.
+
+## Motivation
+
+Script sizes pose a significant problem. This manifests itself in two ways:
+1. Every time a script is used, the transaction which caused the usage must supply the whole script as part of the transaction. This bloats the chain, and passes on the cost of that bloat to users in the form of transaction size fees.
+2. Transaction size limits become are for users. Even if individual scripts do not hit the limits, using multiple scripts has a proportionally greater risk of hitting the limit.
+
+We would like to alleviate these problems.
+
+The key idea is to use reference inputs and modified outputs which carry actual scripts ("reference scripts"), and allow such reference scripts to satisfy the script witnessing requirement for a transaction. 
+This means that the transaction which _uses_ the script will not need to provide it at all, so long as it referenced an output which contained the script.
+
+## Specification
+
+We extend transaction outputs with a new optional field, which contains a script (a "reference script").
+
+The min UTXO value for an output with an additional script field depends on the size of the script, following the `coinsPerUTxOWord` protocol parameter.
+
+When we are validating a transaction and we look for the script corresponding to a script hash, in addition to the scripts provided in the transaction witnesses, we also consider any reference scripts from the outputs referred to by the inputs of the transaction.
+
+### Script context
+
+Scripts are passed information about transactions via the script context.
+We propose to augment the script context to include some information about reference scripts.
+
+Changing the script context will require a new Plutus language version in the ledger to support the new interface.
+The change is: a new optional field is added to outputs and inputs to represent reference scripts.
+Reference scripts are represented by their hash in the script context.
+
+Old versions of the language will retain the old interface.
+We do not propose to try and include information about reference scripts in the old interface.
+
+### CDDL 
+
+The CDDL for transaction outputs will change as follows to reflect the new field.
+```
+transaction_output =
+  [ address
+  , amount : value
+  , ? datum : $hash32
+  , ? ref_script : plutus_script
+  ]
+```
+TODO: can we use a more generic type that allows _any_ script in a forwards-compatible way? 
```

Comment:
You could do something like this:
```
any_script = [ 1, timelock_scipt ] / [ 2, plutus_script ] ;    ... / [ n, other_script ]
```
Our future selves would also be grateful if the tags matched the ones used for the script buckets in either the auxiliary data or the witness sets (which don't match each other 😓)

Reply:
Why not just allow attaching arbitrary data to the UTXO? This would allow it to be used for other purposes too, besides scripts. It does overlap a bit with having inline datums, but I think it's important to consider that the datum might also contain information necessary only for consuming it, and not relevant for the referencer. By having this extended to arbitrary data, we could choose to ignore the datum (which would contain irrelevant information) and only use the information we are interested in.

Reply:
@redxaxder I'd be very happy to just use a production for a discriminated union of scripts if we wanted to promote one (of the two that we apparently have...). Let me know which one!

@L-as The "data" here is used for a specific purpose in the ledger. Just having the field contain arbitrary data has the same issue as putting it in the datum: we then need rules for when it "counts" as a reference script for the purposes of the ledger, and when it's "just" uninterpreted data. Having a field specifically for this purpose avoids any confusion.

On a practical note, the issue of functional confusion between referencing and spending is much less problematic for datums. You can just put a pair in the datum and only look at the first component when spending and the second when referencing. A trick like this isn't viable for reference scripts themselves, where we need the serialized data to exactly match the expected hash.

Reply:
I suggest using the same indices as the auxiliary data, since they're not interleaved with other stuff

Reply:
@michaelpj I don't see why you'd need to know when to "count" it as a reference script or uninterpreted data. Were it arbitrary data, to get the data for a hash specified in a transaction, you would in addition to the attached scripts, `txInfoData`, etc., also check the "reference data" in each input and output. Perhaps there's an issue with this i'm missing?

Reply:
What's the type of the `extra_txout_data` field? 

Is it `Data`? Well, scripts are an incompatible format with `Data`, so we'd have to e.g. parse the `Data` and look for a nested bytestring or something.

Is it just a raw `ByteString`? Now we have an inelegant situation where we just try to match it up with various kinds of thing, and use that to decide how to deserialise it. Which is quite tricky and also not how we treat any other part of the transaction.

I think you're suggesting the latter?

Reply:
A raw bytestring would also be much less useful if the output contains data that isn't supposed to be a hash pre-image for something in the transaction, and just contains extra data you might want to look at. Scripts would then have no easy way of interpreting it (unless we provided `parseData :: ByteString -> Data`, which I'd rather not).

Reply:
You're right, I agree.

Reply:
Using elements on a list of flexible size has always created problems down the road for me. You always need to check length size and access context by the index, even if you use a data structure to parse.
I would already consider for the third entry to be either the datumhash or a map where I can name the content, either datumhash, the inline datum, a timelockscript, a plutus script, and whatever else might come in the future. This lets you put any data, and if you need have a clear way to implement a data parser as the key on the map would specify the parser.